pkg/logs/internal/launchers/file/provider/file_provider.go

```diff
@@ -0,0 +1,324 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package fileprovider
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"sort"
+	"time"
+
+	tailer "github.com/DataDog/datadog-agent/pkg/logs/internal/tailers/file"
+	"github.com/DataDog/datadog-agent/pkg/logs/sources"
+	"github.com/DataDog/datadog-agent/pkg/logs/status"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+
+	"github.com/DataDog/datadog-agent/pkg/logs/config"
+)
+
+// OpenFilesLimitWarningType is the key of the message generated when too many
+// files are tailed
+const openFilesLimitWarningType = "open_files_limit_warning"
+
+// WildcardSelectionStrategy is used to specify if wildcard matches should be prioritized
+// based on their filename or the modification time of each file
+type WildcardSelectionStrategy int
+
+const (
+	// WildcardUseFileModTime means that the top 'filesLimit' most recently modified files
+	// will be chosen from all wildcard matches
+	WildcardUseFileModTime = iota
+	// WildcardUseFileName means that wildcard matches will be chosen in a roughly reverse
+	// lexicographical order
+	WildcardUseFileName
+)
+
+// wildcardOrdering controls what ordering is applied to wildcard matches
+type wildcardOrdering int
+
+const (
+	// wildcardReverseLexicographical is the default option and does a pseudo reverse alpha sort
+	wildcardReverseLexicographical wildcardOrdering = iota
+	// wildcardModTime sorts based on the most recently modified time for each matching wildcard file
+	wildcardModTime
+)
+
+// selectionStrategy controls how the `filesLimit` slots we have are filled given a list of sources
+type selectionStrategy int
+
+const (
+	// greedySelection will consider each source one-by-one, filling as many
+	// slots as is possible from that source before proceeding to the next one
+	greedySelection selectionStrategy = iota
+	// globalSelection will consider files from all sources together and will choose the
+	// top `filesLimit` files based on the `sortMode` ordering
+	globalSelection
+)
+
+// FileProvider implements the logic to retrieve at most filesLimit Files defined in sources
+type FileProvider struct {
+	filesLimit      int
+	wildcardOrder   wildcardOrdering
+	selectionMode   selectionStrategy
+	shouldLogErrors bool
+}
+
+// NewFileProvider returns a new Provider
+func NewFileProvider(filesLimit int, wildcardSelection WildcardSelectionStrategy) *FileProvider {
+	wildcardOrder := wildcardReverseLexicographical
+	selectionMode := greedySelection
```

Comment:
A very minor nit, but would it make sense to remove this at all and use `wildcardSelection` instead? This would allow `applyOrdering` to be split into two functions (it is already called from inside mode-specific branches), allow better name and will make tests simpler (it wouldn't need a fileprovider instance).

Reply:
Happy to split `applyOrdering` into 2 functions, which would indeed make the tests simpler by removing the need for a fileprovider.

To me, separating the state into `wildcardOrder` and `selectionMode` is useful for readability, specifically the `global` vs `greedy` distinction in `FilesToTail`.